Evolving from dependency resolution to dependency management.

jmchilton opened the pull request:
- f597b08 Small toolbox refactor to enable the rest of this (same commit appears in #1398).
- 204aa96 Implement some API endpoints for existing dependency resolution stuff - reload dependency resolves, list resolvers, query to resolve dependency.
- b167a74 Include more information in Dependency object, in particular whether a resolved dependency is `exact` or not - i.e. uses version information.
- 71a8768 Introduce the concept of a ListableDependencyResolver and matching API endpoint. Currently Galaxy packages and Conda dependency resolution are "listable". In theory, module resolver for instance could be listed.
- 29f20cf Introduce the concept of an InstallableDependencyResolver and matching API endpoint. So far only Conda dependency resolver is installable.

Dependency management so simple, so intuitive, so decoupled from how tools are installed and configured. Hopefully it is becoming clear that this should be the future of dependency management.

jmchilton committed:
Refactor ToolBox functionality.

This moves in two directions toward a better design. It moves some Galaxy tool specific logic out of the AbstractToolBox (which should just be worried about panels, etc... IMO) and it moves more galaxy-lib functionality (dependency resolution) into a galaxy-lib module (galaxy.tools.toolbox). It does this by creating a BaseGalaxyToolBox that can reason about which files look like tools and what dependency resoution should be - without dependening on Galaxy models, job running, etc... that is in galaxy.tools.__init__.

jmchilton committed:
API-ify tool dependency resolution.

jmchilton committed:
Include information about whether matches are dependencies are exact.

jmchilton committed:
Implement concept of listable dependency resolvers.

DependencyResolver classes which can enumerate the requirements then can resolve "exactly" should mix in ListableDependencyResolver and implement list_dependencies.

The galaxy package resolver can easily be enumerated but its subclasses like the tool shed dependency resolver couldn't, so I have separated that class into a base class that there others can inherit and an actual resolver that implements ListableDependencyResolver.

Rebase into listable.

jmchilton committed:
Implement concept of installable dependency resolvers.

DependencyResolver classes which can install "simple" dependences should mix in the new InstallableDependencyResolver and implement install_dependency.

Currently this is only the conda dependency resolver, one can easily imagine doing the same for the brew dependency resolver and potentially doing it for some variant of the unlinked dependency resolvers.

bgruening commented on lib/galaxy/tools/deps/__init__.py:
```diff
@@ -107,10 +108,17 @@ def uses_tool_shed_dependencies(self):
             log.debug('Resolver %s returned %s (isnull? %s)' % (resolver.resolver_type, dependency,
                                                                 dependency == INDETERMINATE_DEPENDENCY))
+            if require_exact and not dependency.exact:
+                dependency = INDETERMINATE_DEPENDENCY
```
if/elif?

jmchilton replied:
No, this looks to be what I meant. I want to check that second if condition, regardless of the first condition.

jmchilton committed:
Modify unmodified conda behavior.

 - Add configuration option for overriding condarc location used by conda resolve commands.
 - Default location of condarc to be in tool_dependency_dir.
 - Delay updating the condarc to ensure channels until this would actually be needed (during package installation).
 - Add configuration flag to allow Galaxy to *NOT* check the PATH for conda.
 - Update unit test that verifies conda's installation to never use conda on PATH, since a precondition of this test is that certain packages haven't been installed previously.
 - Change default behavior to not check the PATH for conda, everything else here I feel pretty good about - this I'm ambivalent about - it will reduce error reports and problems with network filesystems, but doesn't seem to be good Unix best practice.

Points 2 and 3 above should address this issue: https://github.com/galaxyproject/ansible-galaxy-extras/pull/33#issuecomment-168384555

This commit also addresses most of #1406.

jmchilton commented:
Just included a bunch of additional conda updates with this:
- Add configuration option for overriding condarc location used by conda resolve commands.
- Default location of condarc to be in tool_dependency_dir.
- Delay updating the condarc to ensure channels until this would actually be needed (during package installation).
- Add configuration flag to allow Galaxy to _NOT_ check the PATH for conda.
- Update unit test that verifies conda's installation to never use conda on PATH, since a precondition of this test is that certain packages haven't been installed previously.
- Change default behavior to not check the PATH for conda, everything else here I feel pretty good about - this I'm ambivalent about - it will reduce error reports and problems with network filesystems, but doesn't seem to be good Unix best practice.

Points 2 and 3 above should address this issue: galaxyproject/ansible-galaxy-extras#33 (comment)

This commit also addresses most of #1406.

bgruening commented:
I will review this today in the evening. Thanks John!

jmchilton committed:
Fixup dependency module and API comments.

Thanks to suggestions by @bgruening.

jmchilton committed:
Fix listing conda requirements if conda hasn't been installed.

Fix problem discovered by @bgruening.

bgruening commented:
Calling http://localhost:8080/api/dependency_resolvers/requirements/ will result in:

```
Traceback (most recent call last):
  File "/home/bag/projects/code/galaxy/lib/galaxy/web/framework/decorators.py", line 258, in decorator
    rval = func( self, trans, *args, **kwargs)
  File "/home/bag/projects/code/galaxy/lib/galaxy/web/framework/decorators.py", line 90, in decorator
    return func( self, trans, *args, **kwargs )
  File "/home/bag/projects/code/galaxy/lib/galaxy/webapps/galaxy/api/tool_dependencies.py", line 135, in manager_requirements
    return self._view.manager_requirements()
  File "/home/bag/projects/code/galaxy/lib/galaxy/tools/deps/views.py", line 29, in manager_requirements
    for requirement in resolver.list_dependencies():
  File "/home/bag/projects/code/galaxy/lib/galaxy/tools/deps/resolvers/conda.py", line 145, in list_dependencies
    for install_target in installed_conda_targets(self.conda_context):
  File "/home/bag/projects/code/galaxy/lib/galaxy/tools/deps/conda_util.py", line 191, in installed_conda_targets
    for name in os.listdir(envs_path):
OSError: [Errno 2] Datei oder Verzeichnis nicht gefunden: '/home/bag/projects/code/galaxy/tool_deps/_conda/envs'
```

This is because `conda_auto_init = False` but should result in a proper error message.

bgruening commented on lib/galaxy/tools/deps/resolvers/__init__.py:
```diff
@@ -1,7 +1,17 @@
+    dict_collection_visible_keys = ['resolver_type', 'resolves_simple_dependencies']
+    # A "simple" dependency is one that does not depend on the the tool
+    # resolving the dependency. Classic tool shed dependencies are non-simple
+    # because the tool shed install.
```
`because the tool shed install`? Because the tool shed needs to resolve the dependencies?

bgruening commented on lib/galaxy/tools/deps/resolvers/galaxy_packages.py:
```diff
@@ -25,40 +31,68 @@ def resolve( self, name, version, type, **kwds ):
+
+
+def _is_dependency_directory(directory):
+    return exists(join(directory, 'env.sh')) or exists(join(directory, 'bin'))
```
join(directory, 'bin') seems to be too relaxed for me. /bin is only recognised as dependency if the parent directory is a symlink named `default` isn't it?

jmchilton replied:
```
    def _find_dep_versioned( self, name, version, type='package', **kwds ):
        base_path = self.base_path
        path = join( base_path, name, version )
        return self._galaxy_package_dep(path, version, True)

    def _find_dep_default( self, name, type='package', exact=True, **kwds ):
        base_path = self.base_path
        path = join( base_path, name, 'default' )
        if islink( path ):
            real_path = realpath( path )
            real_version = basename( real_path )
            return self._galaxy_package_dep(real_path, real_version, exact)
        else:
            return INDETERMINATE_DEPENDENCY

    def _galaxy_package_dep( self, path, version, exact ):
        script = join( path, 'env.sh' )
        if exists( script ):
            return GalaxyPackageDependency(script, path, version, exact)
        elif exists( join( path, 'bin' ) ):
            return GalaxyPackageDependency(None, path, version, exact)
        return INDETERMINATE_DEPENDENCY
```

I don't think this is true, versioned package directories can contain `bin` directories not just symbolically linked default directories. Both the list procedure and resolution method could be updated to check that bin is a directory - but right now they mirror each other which is good enough I think.

bgruening replied:
I was not talking about `versioned packages`, this was more a general comment. I didn't know that a `tool_deps`/`package_name`/`random`/`bin` will be picked up as dependency. Great if it does!

jmchilton replied:
It would get picked up for a requirement with `version="random"`.

bgruening commented on lib/galaxy/webapps/galaxy/api/tool_dependencies.py:
```diff
@@ -0,0 +1,135 @@
+        the match with highest priority.
+
+        :type   index:    int
+        :param  index:    index of the dependncy resolver
```
`dependency`

bgruening commented on lib/galaxy/webapps/galaxy/api/tool_dependencies.py:
```diff
@@ -0,0 +1,135 @@
+        Resolve described requirement against specified dependency resolver.
+
+        :type   index:    int
+        :param  index:    index of the dependncy resolver
```
dependency

bgruening commented on lib/galaxy/webapps/galaxy/api/tool_dependencies.py:
```diff
@@ -0,0 +1,135 @@
+        :type   version:  str
+        :param  version:  version of the requirement to find a dependency for (required)
+        :type   exact:    bool
+        :param  version:  require an exact match to specify requirement (do not discard
```
param exact?

bgruening commented on lib/galaxy/webapps/galaxy/api/tool_dependencies.py:
```diff
@@ -0,0 +1,135 @@
+        :type   version:  str
+        :param  version:  version of the requirement to find a dependency for (required)
+        :type   exact:    bool
+        :param  version:  require an exact match to specify requirement (do not discard
```
param exact?

bgruening commented on lib/galaxy/webapps/galaxy/api/tool_dependencies.py:
```diff
@@ -0,0 +1,135 @@
+                          version information to resolve dependency).
+
+        :rtype:     dict
+        :returns:   a dictified description of the dependency, with type: null
```
with dependency_type: null

bgruening commented on lib/galaxy/webapps/galaxy/api/tool_dependencies.py:
```diff
@@ -0,0 +1,135 @@
+        by all dependency resolvers that support this operation.
+
+        :type   index:    int
+        :param  index:    index of the dependncy resolver
```
dependncy -> dependency

bgruening commented:
Ready to merge if you are :)

jmchilton commented:
I believe I have addressed most of your concerns with ef9e187 and 4650d77 except for https://github.com/galaxyproject/galaxy/pull/1409#discussion_r48693850 which I think is incorrect and what I have included with the PR is fine.

jmchilton commented:
@bgruening Definitely ready for a merge - I have more commits to open a PR based on this.

bgruening commented:
Great thanks John!
